packages/compass-connection-import-export/src/hooks/use-import.ts

```diff
@@ -0,0 +1,156 @@
+};
+
+const INITIAL_STATE: Readonly<ImportConnectionsState> = Object.freeze({
+  ...COMMON_INITIAL_STATE,
+  passphraseRequired: false,
+  fileContents: '',
+});
+
+async function loadFile(
+  {
+    filename,
+    passphrase,
+  }: Pick<ImportConnectionsState, 'filename' | 'passphrase'>,
+  importConnections: typeof dataServiceImportConnections
+): Promise<Partial<ImportConnectionsState>> {
+  if (!filename) {
+    return INITIAL_STATE;
+  }
+  try {
+    const fileContents = await fs.readFile(filename, 'utf8');
+    const connectionList: ConnectionShortInfo[] = [];
+    await importConnections(fileContents, {
+      passphrase,
+      saveConnections(list: ConnectionInfo[]) {
+        for (const info of list) {
+          if (info.favorite?.name) {
+            connectionList.push({
+              name: info.favorite.name,
+              id: info.id,
+              selected: true,
+            });
+          }
+        }
+      },
+    });
+    if (connectionList.length === 0) {
+      throw new Error('File does not contain any connections');
+    }
+    return {
+      fileContents,
+      connectionList,
+      error: '',
+    };
+  } catch (err: any) {
+    return {
+      fileContents: '',
+      connectionList: [],
+      error: err.message,
+      ...(err.passphraseRequired && { passphraseRequired: true }),
+    };
+  }
+}
+
+export function useImportConnections(
+  {
+    finish,
+    open,
+    trackingProps,
+  }: {
+    finish: (result: ImportExportResult) => void;
+    open: boolean;
+    trackingProps?: Record<string, unknown>;
+  },
+  importConnections = dataServiceImportConnections
+): {
+  onCancel: () => void;
+  onSubmit: () => void;
+  onChangeFilename: (filename: string) => void;
+  onChangePassphrase: (passphrase: string) => void;
+  onChangeConnectionList: (connectionInfos: ConnectionShortInfo[]) => void;
+  state: ImportConnectionsState;
+} {
+  const [state, setState] = useState<ImportConnectionsState>(INITIAL_STATE);
+  useEffect(() => setState(INITIAL_STATE), [open]);
+  const { passphrase, filename, fileContents, connectionList } = state;
+
+  const { onChangeConnectionList, onChangePassphrase, onCancel } =
+    useImportExportConnectionsCommon(setState, finish);
+
+  const onSubmit = useCallback(() => {
+    setState((prevState) => ({ ...prevState, inProgress: true }));
+    void (async () => {
+      const filter = makeConnectionInfoFilter(connectionList);
+      try {
+        await importConnections(fileContents, {
+          passphrase,
+          filter,
+          trackingProps,
+        });
+      } catch (err: any) {
+        setState((prevState) => {
+          return {
+            ...prevState,
+            inProgress: false,
+            error: err.message,
+          };
+        });
+        return;
+      }
+      finish('succeeded');
+    })();
+  }, [connectionList, fileContents, passphrase, finish]);
+
+  useEffect(() => {
+    void loadFile({ filename, passphrase }, importConnections).then(
```

Comment:
This probably needs to be debounced and/or `importConnections` needs to be extended with the ability to take an `AbortSignal` or similar to cancel in-flight operations